### PR TITLE
Add correctness-first Domino support to DFlash V2

### PR DESCRIPTION
**Pull request by jianuo-huang**
## Summary

This PR adds Domino projector support to the current DFlash V2 worker for the
public `Huang2020/Qwen3-8B-Domino-b16` checkpoint. It reuses SGLang's existing
target
verification, continuous-prefix acceptance, bonus-token handling, and KV-cache
commit paths unchanged.

The implementation is independent of the earlier integration in #28998 and is
based on the public Domino reference and the current DFlash V2 interfaces.
Addresses #28977 and #29511.

## Design

- Parse, validate, instantiate, and strictly load the checkpoint-native
  bias-free prefix GRU and two-layer SiLU correction projector.
- Generate 15 proposals for the public block-size-16 checkpoint. The first is
  the full-vocabulary DFlash argmax; the GRU is initialized with
  `[verified_token, first_proposal]`, and every realized proposal is fed back
  before the next correction step.
- Keep the base LM-head projection full-vocabulary. Correction scoring uses one
  per-request, block-shared candidate pool with `K=2048` by default; set
  `--speculative-domino-candidate-pool-size 0` for full-vocabulary correction.
- Use native `nn.GRU` for the two-token prefix and ATen `GRUCell` for the 13
  one-token feedback updates.
- Capture the complete rollout in the existing DFlash draft CUDA Graph, with an
  eager fallback when the graph is disabled or unavailable.

The four performance changes are documented separately:

1. [Perf 1: step-major base-logit precompute](https://github.com/sgl-project/sglang/pull/31328#issuecomment-5031396465)
2. [Perf 2: full rollout CUDA Graph](https://github.com/sgl-project/sglang/pull/31328#issuecomment-5035209078)
3. [Perf 3: GRUCell feedback](https://github.com/sgl-project/sglang/pull/31328#issuecomment-5035756608)
4. [Perf 4: block-shared K2048 candidate pool](https://github.com/sgl-project/sglang/pull/31328#issuecomment-5035759488)

## Supported and numerical scope

- Tested scope: A100 80GB, CUDA, BF16, TP=1 for Qwen3-8B,
  `pure_draft_prefix_len=1`, and block size 16.
- Graph results cover exact capture batch shapes with
  `--disable-cuda-graph-padding`. Padded replay is not claimed.
- ATen GRUCell and cuDNN GRU use different BF16 operation orderings, so their
  proposal tokens are not generally bit-exact.
- `K=2048` is an approximate search for the correction argmax. Target
  verification is unchanged, but the observed K2048-versus-K0 output parity is
  limited to the fixed workloads reported below; it is not a general
  equivalence guarantee.

## Validation

- Registered eager GSM8K-200 gate with public models/data: score `0.950`, average
  speculative acceptance length `4.5544` (gates: `>=0.90` and `>4.0`).
- Exact-shape Perf 1 versus Perf 2 graph A/B: GSM8K-200 score `0.950` in both
  arms and `0/208` fixed-512 ShareGPT text mismatches.
- Perf 3 eager versus exact-shape graph replay: `0/21,900` proposal mismatches
  over 20 random trials per batch. GSM8K-200 remained `0.950`.
- Perf 4 K2048 versus full-vocabulary K0: GSM8K-32 matched `32/32` output-ID
  sequences with acceptance length `4.7975`; ShareGPT C32 matched `128/128`
  texts with acceptance length `3.7980`.
- Focused config, weight-loading, runtime validation, full-chain, batch
  independence, candidate-pool, sampler, and CUDA Graph replay tests:
  `22 passed`, `23 subtests passed`.

```bash
CUDA_VISIBLE_DEVICES=0 PYTHONPATH=python \
python3 -m pytest -q -s \
  test/registered/spec/dflash/test_dflash_domino.py

PYTHONPATH=python python3 -m pytest -q \
  test/registered/unit/spec/test_dflash_domino.py
```

## Performance summary

The first table is a standalone Domino-rollout benchmark on one A100 80GB,
BF16/TP=1, 15 proposals, exact batch shapes, and full-vocabulary correction
(`K=0`). Target verification and the DFlash backbone are excluded. Values are
the median of six balanced measurements with 30 CUDA-event iterations each.

| Batch | No optimization | Perf 1 | Perf 2 | Perf 3 | No-opt to Perf 3 |
|---:|---:|---:|---:|---:|---:|
| 1 | 13.416 ms | 5.888 ms | 3.372 ms | 2.363 ms *(base: 0.674 ms)* | 5.68x |
| 8 | 14.104 ms | 6.023 ms | 3.875 ms | 2.881 ms *(base: 0.775 ms)* | 4.90x |
| 64 | 15.851 ms | 8.821 ms | 8.571 ms | 7.808 ms *(base: 4.555 ms)* | 2.03x |

Parenthesized base-logit time is included in the total but belongs to DFlash; it
is not Domino-specific overhead.

Perf 4 uses a separate candidate-pool harness and aggregation, so it is reported
as its own K0 A/B rather than appended to the table above.

| Batch | Full correction, K0 | Candidate pool, K2048 | Reduction |
|---:|---:|---:|---:|
| 1 | 2.441 ms *(base: 0.724 ms)* | 1.618 ms *(base: 0.724 ms)* | 33.7% |
| 8 | 2.879 ms *(base: 0.817 ms)* | 2.122 ms *(base: 0.817 ms)* | 26.3% |
| 64 | 7.700 ms *(base: 5.104 ms)* | 7.188 ms *(base: 5.104 ms)* | 6.7% |

The same parenthesized DFlash base-logit time is included in both Perf 4 arms.

Using SGLang's official serving benchmark runner:

- Perf 2 versus Perf 1 improved output throughput by `11.2%`, `3.4%`, and
  `1.5%` at concurrency 1/8/32.
- Perf 3 versus Perf 2 improved output throughput by `5.8%`, `3.5%`, and
  `1.9%`; C32 acceptance length changed by `-0.16%`.
- Perf 4 K2048 versus K0 measured `1450.31` versus `1431.25` output tok/s at
  C32 (`+1.3%`). This was one run per K and is treated as exploratory/noise-level,
  not a serving-speedup claim.

### Final end-to-end serving benchmark

A final 135-cell TP1 serving sweep compared target-only decoding, the official
Qwen3-8B DFlash b16 draft, and Domino b16 K2048 on one A100 80GB.

[Qwen3-8B end-to-end serving throughput](https://github.com/sgl-project/sglang/pull/31328#issuecomment-5041901375)

Full settings and all numerical results: [benchmark comment](https://github.com/sgl-project/sglang/pull/31328#issuecomment-5041901375).

Rollout microbenchmark speedups are not end-to-end serving claims. Exact setup,
correctness data, caveats, figures, and benchmark commands are in the four
linked performance comments.

## Checklist

- [x] Format changed files with the repository pre-commit hooks.
- [x] Add focused CPU/CUDA unit tests and a registered CUDA E2E gate.
- [x] Provide accuracy, acceptance-length, rollout, and official serving results.
- [ ] Add user documentation after the supported runtime contract is agreed.

---
### CI States

Latest PR Test (Base): :x: [Run #33199855245](https://github.com/sgl-project/sglang/actions/runs/33199855245)
Latest PR Test (Extra): :x: [Run #33199855055](https://github.com/sgl-project/sglang/actions/runs/33199855055)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33199855228](https://github.com/sgl-project/sglang/actions/runs/33199855228)

### CODE REVIEW

File: python/sglang/srt/speculative/domino_utils.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        step_hidden = z[:, index, :]
+        correction = embed_proj(torch.cat((step_hidden, gru_hidden[0]), dim=-1))
+        next_ids = torch.argmax(base_logits(step_hidden) + correction, dim=-1).to(
+            torch.long
+        )
+        proposals.append(next_ids)
+        if index + 1 < num_proposals:
+            _, gru_hidden = prefix_gru(target_embedding(next_ids[:, None]), gru_hidden)
+
+    return torch.stack(proposals, dim=1)
```
[the remark this answers is not in the document]
Comment (jianuo-huang) [reply python/sglang/srt/speculative/domino_utils.py:198]: Addressed in `a7bb8af1cb`. A direct `[B, P, V]` layout regressed at B=64 because each per-step slice was non-contiguous, so this uses one step-major LM-head GEMM with contiguous `[B, V]` slices. Full-rollout A100 medians improved from 12.32 to 3.81 ms (B=1), 12.96 to 3.81 ms (B=8), and 14.76 to 7.75 ms (B=64), with exact token parity in all cases.
